.config/doom/config.org
#+TITLE: Emacs Configuration
#+STARTUP: overview
#+PROPERTY: header-args:emacs-lisp      :tangle "./config.el" :results silent

* Startup Performance
Make startup faster by reducing the frequency of garbage collection and then use a hook to measure Emacs starup time.
#+begin_src emacs-lisp
;; The default is 800 kilobytes. Measured in bytes.
(setq gc-cons-threshold (* 500 1024 1024)) ;; 500mb

;;Increase the amount of data which Emacs reads from the process
(setq read-process-output-max (* 5 1024 1024)) ;; 5mb

;; Profile emacs startup
(add-hook 'emacs-startup-hook
          (lambda ()
            (message "*** Emacs loaded in %s with %d garbage collections"
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))
#+end_src
* Completion
#+begin_src emacs-lisp
(setq completion-ignore-case t)
#+end_src
* Doom
** Theme
There are two ways to load a theme. Both assume the theme is installed and available. You can either set `doom-theme' or manually load a theme with the `load-theme' function.
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-one)
;; (setq catppuccin-flavor 'mocha)
;; (setq doom-theme 'catppuccin)
#+END_SRC
** Font
Doom exposes five (optional) variables for controlling fonts in Doom. Here are the three important ones:
- `doom-font'
- `doom-variable-pitch-font'
- `doom-big-font' -- used for `doom-big-font-mode'; use this for presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd font string. You generally only need these two:
(setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light) doom-variable-pitch-font (font-spec :family "sans" :size 13))
#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Hack Nerd Font Mono" :size 18))
#+END_SRC
** Title
Include project and file/buffer name in the title
#+begin_src emacs-lisp
(setq-default frame-title-format
              '((:eval (projectile-project-name)) " - %b"))
#+end_src
** Line Numbers
This determines the style of line numbers in effect. If set to `nil', line numbers are disabled. For relative line numbers, set this to `relative'.
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type 'visual)
#+END_SRC
** Additional functions/macros
Here are some additional functions/macros that could help you configure Doom:
- `load!' for loading external *.el files relative to this one
- `use-package!' for configuring packages
- `after!' for running code after a package has loaded
- `add-load-path!' for adding directories to the `load-path', relative to this file. Emacs searches the `load-path' when you load packages with `require' or `use-package'.
- `map!' for binding new keys

To get information about any of these functions/macros, move the cursor over the highlighted symbol at press 'K' (non-evil users must press 'C-c c k'). This will open documentation for it, including demos of how they are used.

You can also try 'gd' (or 'C-c c d') to jump to their definition and see how they are implemented.
* Window
Persist Emacs' initial frame position, dimension and/or full-screen state across sessions
#+BEGIN_SRC emacs-lisp
(when-let (dims (doom-store-get 'last-frame-size))
  (cl-destructuring-bind ((left . top) width height fullscreen) dims
    (setq initial-frame-alist
          (append initial-frame-alist
                  `((left . ,left)
                    (top . ,top)
                    (width . ,width)
                    (height . ,height)
                    (fullscreen . ,fullscreen))))))

(defun save-frame-dimensions ()
  (doom-store-put 'last-frame-size
                  (list (frame-position)
                        (frame-width)
                        (frame-height)
                        (frame-parameter nil 'fullscreen))))

(add-hook 'kill-emacs-hook #'save-frame-dimensions)
#+END_SRC
* Ivy
Use fuzzy search. (using ivy +fuzzy in init.el)
** Prescient
#+begin_src emacs-lisp
(setq prescient-history-length 0)
(setq prescient-filter-method '(anchored))
(setq ivy-prescient-enable-filtering t)
(setq ivy-prescient-enable-sorting t)
#+end_src

* Major-Modes
Remap major modes to tree-sitter
#+begin_src elisp
(setq major-mode-remap-alist
      '((css-mode . css-ts-mode)
        (rust-mode . rust-ts-mode)))
#+end_src
* Org mode
** Directory
If you use `org' and don't want your org files in the default location below, change `org-directory'. It must be set before org loads!
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/org/")
(setq org-archive-location "~/org/archive")
#+END_SRC

** Capture templates
#+BEGIN_SRC emacs-lisp
(after! org (setq org-capture-templates
      '(("t" "Todo" entry
         (file+headline "~/org/refile.org" "Inbox")
         "* TODO %^{Task}
:PROPERTIES:
:CREATED: %U
:END:
%?")

        ("e" "Event" entry
         (file+headline "~/org/calendar.org" "Events")
         "* %^{Event}\n%^{SCHEDULED}T\n:PROPERTIES:\n:CREATED: %U\n:CAPTURED: %a\n:CONTACT: %(org-capture-ref-link \"~/org/contacts.org\")\n:END:\n%?")

        ("c" "Contact" entry
         (file "~/org/contacts.org")
         "* %^{Name}
:PROPERTIES:
:CREATED: %U
:EMAIL: %^{Email}
:PHONE: %^{Phone}
:END:
\\ *** Communications
\\ *** Notes
%?"))))

;; Helper function to select and link a contact
(defun org-capture-ref-link (file)
  "Create a link to a contact in contacts.org"
  (let* ((headlines (org-map-entries
                     (lambda ()
                       (cons (org-get-heading t t t t)
                             (org-id-get-create)))
                     t
                     (list file)))
         (contact (completing-read "Contact: "
                                   (mapcar #'car headlines)))
         (id (cdr (assoc contact headlines))))
    (format "[[id:%s][%s]]" id contact)))
#+END_SRC

** When done
Add timestamp and note when done
#+BEGIN_SRC emacs-lisp
(setq org-log-done 'time)
#+END_SRC
** Clocktables
*** Only print hours and minuts for time
#+begin_src emacs-lisp
(setq org-duration-format (quote h:mm))
#+end_src
** MPV Playback
#+begin_src emacs-lisp
(defun tm-open-with-mpv (path _prefix)
  (let ((file (expand-file-name path)))
    (make-process
     :name "mpv"
     :command (list "mpv" file))))

(after! org
  (org-link-set-parameters "mpv" :follow #'tm-open-with-mpv))
#+end_src
* Keyboard
** Ripgrep
#+begin_src emacs-lisp
(map! :leader
      (:prefix ("r" . "ripgrep")
       :desc "Project" "r" #'rg-project))
#+end_src
** Lazygit
#+begin_src emacs-lisp
(map! :leader
      :desc "Lazygit" "l" #'tm:lazygit)
#+end_src
* LSP
#+begin_src emacs-lisp
(setq lsp-auto-execute-action nil)
(setq lsp-signature-auto-activate nil)

(setq lsp-javascript-preferences-import-module-specifier "non-relative")
(setq lsp-typescript-preferences-import-module-specifier "non-relative")
#+end_src
** QML
#+begin_src emacs-lisp
(when (require 'treesit nil t)
  (add-to-list 'treesit-language-source-alist
               '(qmljs "https://github.com/yuja/tree-sitter-qmljs")))
  ;; (add-to-list 'treesit-load-name-override-list
  ;;              '(qmljs "libtree-sitter-qml" "tree_sitter_qmljs")))

(use-package! qml-ts-mode
  :mode "\\.qml\\'"
  :config
  (add-to-list 'major-mode-remap-alist '(qml-mode . qml-ts-mode)))

(after! lsp-mode
  (add-to-list 'lsp-language-id-configuration '(qml-ts-mode . "qml-ts"))
  (lsp-register-client
   (make-lsp-client
    :new-connection (lsp-stdio-connection '("qmlls6"))
    :activation-fn (lsp-activate-on "qml-ts")
    :server-id 'qmls)))

(add-hook! 'qml-ts-mode-hook
           (setq-local electric-indent-chars
                       '(?\n ?\( ?\) ?{ ?} ?\[ ?\] ?\; ?,))
           (lsp-deferred))
#+end_src
*** Quickshell
For Quickshell, create an empty file named .qmlls.ini in the same directory as your shell.qml. Quickshell will replace it with a managed qmlls configuration, and the docs recommend gitignoring it because it is machine-specific:
https://quickshell.org/docs/master/guide/install-setup
https://quickshell.org/docs/v0.2.1/guide/install-setup/#language-server

So if your shell entrypoint is, for example, ~/.config/quickshell/shell.qml, then do:

touch ~/.config/quickshell/.qmlls.ini

A few caveats from the docs:

- qmlls works poorly when the file structure is temporarily broken.
- Quickshell type docs are not available through the LSP.
- PanelWindow may not resolve correctly.
** Odin
#+begin_src emacs-lisp
(use-package! odin-ts-mode
  :mode "\\.odin\\'")

(after! treesit
  (add-to-list 'treesit-language-source-alist
               '(odin "https://github.com/tree-sitter-grammars/tree-sitter-odin")))

(after! lsp-mode
  (add-to-list 'lsp-language-id-configuration '(odin-ts-mode . "odin"))
  (lsp-register-client
   (make-lsp-client
    :new-connection (lsp-stdio-connection '("ols"))
    :activation-fn (lsp-activate-on "odin")
    :server-id 'ols)))

(add-hook! 'odin-ts-mode-hook
  (lsp-deferred))
#+end_src
** Lua
Add hyprland stubs.
#+begin_src emacs-lisp
(after! lsp-lua
  (setq lsp-lua-workspace-library
        (vconcat (or lsp-lua-workspace-library [])
                 ["/usr/share/hypr/stubs"])))
#+end_src
* iBuffer
** Group files
#+begin_src emacs-lisp
(setq ibuffer-expert t)
(setq ibuffer-display-summary nil)
(setq ibuffer-use-other-window nil)
(setq ibuffer-show-empty-filter-groups nil)
(setq ibuffer-default-sorting-mode 'filename/process)
(setq ibuffer-title-face 'font-lock-doc-face)
(setq ibuffer-use-header-line t)
(setq ibuffer-default-shrink-to-minimum-size nil)

(setq ibuffer-saved-filter-groups
      '(("Main"
         ("Directories" (mode . dired-mode))
         ("Org" (mode . org-mode))
         ("Emacs" (or
                     (mode . emacs-lisp-mode)
                     (name . "^\\*Help\\*$")
                     (name . "^\\*Custom.*")
                     (name . "^\\*Org Agenda\\*$")
                     (name . "^\\*info\\*$")
                     (name . "^\\*scratch\\*$")
                     (name . "^\\*Backtrace\\*$")
                     (name . "^\\*doom\\*$")
                     (name . "^\\*compilation\\*")
                     (name . "^\\*Async-native-compile-log\\*$")
                     (name . "^\\*Native-compile-log\\*$")
                     (name . "^\\*Messages\\*$"))))))

(add-hook 'ibuffer-mode-hook (lambda () (ibuffer-switch-to-saved-filter-groups "Main")))
#+end_src
* vertico-posframe
#+begin_src emacs-lisp
;; (vertico-posframe-mode 1)
#+end_src
* Utility Functions
#+begin_src emacs-lisp
(defun tm:open-eshell-at (path &optional name)
  "Open eshell at specified path"
  (interactive)
  (eshell '(4))
  (eshell-return-to-prompt)
  (insert (concat "cd " path))
  (eshell-send-input)
  (if name (rename-buffer name)))

(defun tm:load-config (file)
  "Load config file"
  (interactive)
  (setq path (expand-file-name file "~/.config/doom/"))
  (cond
   ((file-exists-p path)
    (load-file path))
   (t (message "Loading %s... not found" path))))

(defun tm:org-copy-cell-content ()
  "Copy cell content from org table"
  (interactive)
  (when (org-at-table-p)
    (kill-new
     (string-trim
      (substring-no-properties(org-table-get-field))))))

(defun tm:prettier-content ()
  "Format file content using prettier from current project"
  (interactive)
  (shell-command-on-region
   ;; beginning and end of buffer
   (point-min)
   (point-max)
   ;; command and parameters
   "npx prettier --stdin-filepath temp.ts"
   ;; output buffer
   (current-buffer)
   ;; replace
   t
   "*prettier error buffer"
   ;; show error buffer
   t)
  (save-buffer))
#+end_src
** Lazygit integration
#+begin_src emacs-lisp
(defun tm:lazygit ()
  "Open lazygit (in alacritty) for the current project root"
  (interactive)
  (start-process
   "run lazygit"
   nil
   "alacritty"
   "-t"
   "lazygit-floating"
   "-e"
   "lazygit"
   "-p"
   (or (doom-project-root) default-directory)))
#+end_src
* Company
** Faster completion
#+begin_src emacs-lisp
(setq company-idle-delay 0.01)
(setq company-minimum-prefix-length 1)
#+end_src
** Use TAB for completion
#+begin_src emacs-lisp
(after! company
  ;;; Prevent suggestions from being triggered automatically. In particular,
  ;;; this makes it so that:
  ;;; - TAB will always complete the current selection.
  ;;; - RET will only complete the current selection if the user has explicitly
  ;;;   interacted with Company.
  ;;; - SPC will never complete the current selection.
  ;;;
  ;;; Based on:
  ;;; - https://github.com/company-mode/company-mode/issues/530#issuecomment-226566961
  ;;; - https://emacs.stackexchange.com/a/13290/12534
  ;;; - http://stackoverflow.com/a/22863701/3538165
  ;;;
  ;;; See also:
  ;;; - https://emacs.stackexchange.com/a/24800/12534
  ;;; - https://emacs.stackexchange.com/q/27459/12534

  ;; <return> is for windowed Emacs; RET is for terminal Emacs
  (dolist (key '("<return>" "RET"))
    ;; Here we are using an advanced feature of define-key that lets
    ;; us pass an "extended menu item" instead of an interactive
    ;; function. Doing this allows RET to regain its usual
    ;; functionality when the user has not explicitly interacted with
    ;; Company.
    (define-key company-active-map (kbd key)
      `(menu-item nil company-complete
                  :filter ,(lambda (cmd)
                             (when (company-explicit-action-p)
                              cmd)))))
  ;; (define-key company-active-map (kbd "TAB") #'company-complete-selection)
  (map! :map company-active-map "TAB" #'company-complete-selection)
  (map! :map company-active-map "<tab>" #'company-complete-selection)
  (define-key company-active-map (kbd "SPC") nil)

  ;; Company appears to override the above keymap based on company-auto-complete-chars.
  ;; Turning it off ensures we have full control.
  (setq company-auto-commit-chars nil)
  )
#+end_src
* Maps
#+begin_src emacs-lisp
(map! :leader
      (:prefix ("c")
       :desc "prettier" "p" (cmd! (tm:prettier-content))))
#+end_src
* Load Private Configuration
Load private/machine specific configuration
#+begin_src emacs-lisp
(tm:load-config "config_private.el")
#+end_src


.config/doom/config.el
;; The default is 800 kilobytes. Measured in bytes.
(setq gc-cons-threshold (* 500 1024 1024)) ;; 500mb

;;Increase the amount of data which Emacs reads from the process
(setq read-process-output-max (* 5 1024 1024)) ;; 5mb

;; Profile emacs startup
(add-hook 'emacs-startup-hook
          (lambda ()
            (message "*** Emacs loaded in %s with %d garbage collections"
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))

(setq completion-ignore-case t)

(setq doom-theme 'doom-one)
;; (setq catppuccin-flavor 'mocha)
;; (setq doom-theme 'catppuccin)

(setq doom-font (font-spec :family "Hack Nerd Font Mono" :size 18))

(setq-default frame-title-format
              '((:eval (projectile-project-name)) " - %b"))

(setq display-line-numbers-type 'visual)

(when-let (dims (doom-store-get 'last-frame-size))
  (cl-destructuring-bind ((left . top) width height fullscreen) dims
    (setq initial-frame-alist
          (append initial-frame-alist
                  `((left . ,left)
                    (top . ,top)
                    (width . ,width)
                    (height . ,height)
                    (fullscreen . ,fullscreen))))))

(defun save-frame-dimensions ()
  (doom-store-put 'last-frame-size
                  (list (frame-position)
                        (frame-width)
                        (frame-height)
                        (frame-parameter nil 'fullscreen))))

(add-hook 'kill-emacs-hook #'save-frame-dimensions)

(setq prescient-history-length 0)
(setq prescient-filter-method '(anchored))
(setq ivy-prescient-enable-filtering t)
(setq ivy-prescient-enable-sorting t)

(setq org-directory "~/org/")
(setq org-archive-location "~/org/archive")

(after! org (setq org-capture-templates
      '(("t" "Todo" entry
         (file+headline "~/org/refile.org" "Inbox")
         "* TODO %^{Task}
:PROPERTIES:
:CREATED: %U
:END:
%?")

        ("e" "Event" entry
         (file+headline "~/org/calendar.org" "Events")
         "* %^{Event}\n%^{SCHEDULED}T\n:PROPERTIES:\n:CREATED: %U\n:CAPTURED: %a\n:CONTACT: %(org-capture-ref-link \"~/org/contacts.org\")\n:END:\n%?")

        ("c" "Contact" entry
         (file "~/org/contacts.org")
         "* %^{Name}
:PROPERTIES:
:CREATED: %U
:EMAIL: %^{Email}
:PHONE: %^{Phone}
:END:
\\ *** Communications
\\ *** Notes
%?"))))

;; Helper function to select and link a contact
(defun org-capture-ref-link (file)
  "Create a link to a contact in contacts.org"
  (let* ((headlines (org-map-entries
                     (lambda ()
                       (cons (org-get-heading t t t t)
                             (org-id-get-create)))
                     t
                     (list file)))
         (contact (completing-read "Contact: "
                                   (mapcar #'car headlines)))
         (id (cdr (assoc contact headlines))))
    (format "[[id:%s][%s]]" id contact)))

(setq org-log-done 'time)

(setq org-duration-format (quote h:mm))

(defun tm-open-with-mpv (path _prefix)
  (let ((file (expand-file-name path)))
    (make-process
     :name "mpv"
     :command (list "mpv" file))))

(after! org
  (org-link-set-parameters "mpv" :follow #'tm-open-with-mpv))

(map! :leader
      (:prefix ("r" . "ripgrep")
       :desc "Project" "r" #'rg-project))

(map! :leader
      :desc "Lazygit" "l" #'tm:lazygit)

(setq lsp-auto-execute-action nil)
(setq lsp-signature-auto-activate nil)

(setq lsp-javascript-preferences-import-module-specifier "non-relative")
(setq lsp-typescript-preferences-import-module-specifier "non-relative")

(when (require 'treesit nil t)
  (add-to-list 'treesit-language-source-alist
               '(qmljs "https://github.com/yuja/tree-sitter-qmljs")))
  ;; (add-to-list 'treesit-load-name-override-list
  ;;              '(qmljs "libtree-sitter-qml" "tree_sitter_qmljs")))

(use-package! qml-ts-mode
  :mode "\\.qml\\'"
  :config
  (add-to-list 'major-mode-remap-alist '(qml-mode . qml-ts-mode)))

(after! lsp-mode
  (add-to-list 'lsp-language-id-configuration '(qml-ts-mode . "qml-ts"))
  (lsp-register-client
   (make-lsp-client
    :new-connection (lsp-stdio-connection '("qmlls6"))
    :activation-fn (lsp-activate-on "qml-ts")
    :server-id 'qmls)))

(add-hook! 'qml-ts-mode-hook
           (setq-local electric-indent-chars
                       '(?\n ?\( ?\) ?{ ?} ?\[ ?\] ?\; ?,))
           (lsp-deferred))

(use-package! odin-ts-mode
  :mode "\\.odin\\'")

(after! treesit
  (add-to-list 'treesit-language-source-alist
               '(odin "https://github.com/tree-sitter-grammars/tree-sitter-odin")))

(after! lsp-mode
  (add-to-list 'lsp-language-id-configuration '(odin-ts-mode . "odin"))
  (lsp-register-client
   (make-lsp-client
    :new-connection (lsp-stdio-connection '("ols"))
    :activation-fn (lsp-activate-on "odin")
    :server-id 'ols)))

(add-hook! 'odin-ts-mode-hook
  (lsp-deferred))

(after! lsp-lua
  (setq lsp-lua-workspace-library
        (vconcat (or lsp-lua-workspace-library [])
                 ["/usr/share/hypr/stubs"])))

(setq ibuffer-expert t)
(setq ibuffer-display-summary nil)
(setq ibuffer-use-other-window nil)
(setq ibuffer-show-empty-filter-groups nil)
(setq ibuffer-default-sorting-mode 'filename/process)
(setq ibuffer-title-face 'font-lock-doc-face)
(setq ibuffer-use-header-line t)
(setq ibuffer-default-shrink-to-minimum-size nil)

(setq ibuffer-saved-filter-groups
      '(("Main"
         ("Directories" (mode . dired-mode))
         ("Org" (mode . org-mode))
         ("Emacs" (or
                     (mode . emacs-lisp-mode)
                     (name . "^\\*Help\\*$")
                     (name . "^\\*Custom.*")
                     (name . "^\\*Org Agenda\\*$")
                     (name . "^\\*info\\*$")
                     (name . "^\\*scratch\\*$")
                     (name . "^\\*Backtrace\\*$")
                     (name . "^\\*doom\\*$")
                     (name . "^\\*compilation\\*")
                     (name . "^\\*Async-native-compile-log\\*$")
                     (name . "^\\*Native-compile-log\\*$")
                     (name . "^\\*Messages\\*$"))))))

(add-hook 'ibuffer-mode-hook (lambda () (ibuffer-switch-to-saved-filter-groups "Main")))

;; (vertico-posframe-mode 1)

(defun tm:open-eshell-at (path &optional name)
  "Open eshell at specified path"
  (interactive)
  (eshell '(4))
  (eshell-return-to-prompt)
  (insert (concat "cd " path))
  (eshell-send-input)
  (if name (rename-buffer name)))

(defun tm:load-config (file)
  "Load config file"
  (interactive)
  (setq path (expand-file-name file "~/.config/doom/"))
  (cond
   ((file-exists-p path)
    (load-file path))
   (t (message "Loading %s... not found" path))))

(defun tm:org-copy-cell-content ()
  "Copy cell content from org table"
  (interactive)
  (when (org-at-table-p)
    (kill-new
     (string-trim
      (substring-no-properties(org-table-get-field))))))

(defun tm:prettier-content ()
  "Format file content using prettier from current project"
  (interactive)
  (shell-command-on-region
   ;; beginning and end of buffer
   (point-min)
   (point-max)
   ;; command and parameters
   "npx prettier --stdin-filepath temp.ts"
   ;; output buffer
   (current-buffer)
   ;; replace
   t
   "*prettier error buffer"
   ;; show error buffer
   t)
  (save-buffer))

(defun tm:lazygit ()
  "Open lazygit (in alacritty) for the current project root"
  (interactive)
  (start-process
   "run lazygit"
   nil
   "alacritty"
   "-t"
   "lazygit-floating"
   "-e"
   "lazygit"
   "-p"
   (or (doom-project-root) default-directory)))

(setq company-idle-delay 0.01)
(setq company-minimum-prefix-length 1)

(after! company
  ;;; Prevent suggestions from being triggered automatically. In particular,
  ;;; this makes it so that:
  ;;; - TAB will always complete the current selection.
  ;;; - RET will only complete the current selection if the user has explicitly
  ;;;   interacted with Company.
  ;;; - SPC will never complete the current selection.
  ;;;
  ;;; Based on:
  ;;; - https://github.com/company-mode/company-mode/issues/530#issuecomment-226566961
  ;;; - https://emacs.stackexchange.com/a/13290/12534
  ;;; - http://stackoverflow.com/a/22863701/3538165
  ;;;
  ;;; See also:
  ;;; - https://emacs.stackexchange.com/a/24800/12534
  ;;; - https://emacs.stackexchange.com/q/27459/12534

  ;; <return> is for windowed Emacs; RET is for terminal Emacs
  (dolist (key '("<return>" "RET"))
    ;; Here we are using an advanced feature of define-key that lets
    ;; us pass an "extended menu item" instead of an interactive
    ;; function. Doing this allows RET to regain its usual
    ;; functionality when the user has not explicitly interacted with
    ;; Company.
    (define-key company-active-map (kbd key)
      `(menu-item nil company-complete
                  :filter ,(lambda (cmd)
                             (when (company-explicit-action-p)
                              cmd)))))
  ;; (define-key company-active-map (kbd "TAB") #'company-complete-selection)
  (map! :map company-active-map "TAB" #'company-complete-selection)
  (map! :map company-active-map "<tab>" #'company-complete-selection)
  (define-key company-active-map (kbd "SPC") nil)

  ;; Company appears to override the above keymap based on company-auto-complete-chars.
  ;; Turning it off ensures we have full control.
  (setq company-auto-commit-chars nil)
  )

(map! :leader
      (:prefix ("c")
       :desc "prettier" "p" (cmd! (tm:prettier-content))))

(tm:load-config "config_private.el")
